data/lua/core/plugin.lua

```diff
@@ -0,0 +1,82 @@
+--[========[Plugin module]========]
+
+---@alias plugin_idle_function fun(ctx_name:string, events:WMLTable):boolean|nil
+
+if wesnoth.kernel_type() == "Application Lua Kernel" then
+	print("Loading plugin module...")
+	wesnoth.plugin = {}
+
+	---Yields control back to the game until the next slice.
+	---@return WMLTable
+	---@return plugin_context
+	---@return plugin_info
+	function wesnoth.plugin.next_slice()
+		return coroutine.yield()
+	end
+
+	---@param cond fun(info:plugin_info):boolean
+	---@param idle plugin_idle_function
+	---@return WMLTable
+	---@return plugin_context
+	---@return plugin_info
+	local function wait_until(cond, idle)
+		local events, context, info = {}, nil, nil
+		repeat
+			local new_events
+			new_events, context, info = wesnoth.plugin.next_slice()
+			for i = 1, #new_events do
+				events[#events + 1] = new_events[i]
+			end
+			if idle then
+				if idle(info.name, events) then
+					break
+				end
+			end
+			if info.name == 'Dialog' then
+				context.skip_dialog{}
+			end
+		until cond(info)
+		return events, context, info
+	end
+
+	---Waits until the plugin reaches a specified context.
+	---Unless the idle function returns true, the context returned from this function is
+	---guaranteed to have the expected value.
+	---@param ctx string The context to wait for.
```

Comment:
Calling this also context seems confusing.

Reply:
Maybe `ctx_name`?

Reply:
It's difficult to understand what the difference between info and context is if we call info.name also context name. Though I guess it's rather that info.name is badly named...

Reply:
I think I'd say that all three elements together (events, context, info) constitute the context, so the variable that's  badly named is context rather than info.

Reply:
Is there a reason why they are all separate? I guess it makes sense for `events` but we could join `info` and `context` into one? Is there a good chance for overlap in names? Is it important for the user to understand the principle difference between context and info? Could also be something like `context.name`, `context.callbacks` and `context.accessors` or so.

Reply:
> Is there a reason why they are all separate?

I have no idea. I've been pondering merging them together too. I don't think it's likely that names would overlap between info and context.